GeneralsMD/Code/Tools/WorldBuilder/src/wbview3d.cpp

```diff
@@ -3253,3 +3270,146 @@ void WbView3d::OnUpdateViewShowSoundCircles(CCmdUI* pCmdUI)
+void WbView3d::OnMouseMove(UINT nFlags, CPoint point)
+{
+	// Call base class first
+	WbView::OnMouseMove(nFlags, point);
+	
+	// Don't invalidate here - let drawLabels handle it naturally
+	// Invalidating here causes flicker due to paint feedback loops
+}
+
+//=============================================================================
+// WbView3d::drawBrushModeHint
+//=============================================================================
+/** Draws the brush mode hint on canvas near the brush cursor in 3D view. */
+//=============================================================================
+void WbView3d::drawBrushModeHint(HDC hdc)
+{
+	if (m_hintDrawnThisFrame) {
+		return;
+	}
+	
+	Tool *pCurTool = WbApp()->getCurTool();
+	if (!pCurTool || pCurTool->getToolID() != ID_BRUSH_TOOL) {
+		if (!::IsRectEmpty(&m_lastHintRect)) {
+			::SetRectEmpty(&m_lastHintRect);
+			m_lastBrushMode = -1;
+			m_lastHintPos = CPoint(-10000, -10000);
+		}
+		return;
+	}
+
+	if (::GetAsyncKeyState(VK_LBUTTON) & 0x8000) {
+		if (!::IsRectEmpty(&m_lastHintRect)) {
+			::SetRectEmpty(&m_lastHintRect);
+			m_lastBrushMode = -1;
+			m_lastHintPos = CPoint(-10000, -10000);
+		}
+		return;
+	}
+
+	BrushTool::EBrushMode currentMode = BrushTool::getPreviewModeFromKeys();
+	Int modeInt = (Int)currentMode;
+
+	Coord3D brushPos = DrawObject::getFeedbackPos();
+	
+	Vector3 worldPos, screenPos;
+	worldPos.Set(brushPos.x, brushPos.y, brushPos.z);
+	if (CameraClass::INSIDE_FRUSTUM != m_camera->Project(screenPos, worldPos)) {
+		if (!::IsRectEmpty(&m_lastHintRect)) {
+			::SetRectEmpty(&m_lastHintRect);
+			m_lastBrushMode = -1;
+			m_lastHintPos = CPoint(-10000, -10000);
+		}
+		return;
+	}
+
+	CRect rClient;
+	GetClientRect(&rClient);
+	Int sx, sy;
+	W3DLogicalScreenToPixelScreen(screenPos.X, screenPos.Y, &sx, &sy,
+		rClient.right - rClient.left, rClient.bottom - rClient.top);
+	CPoint hintPos;
+	hintPos.x = rClient.left + sx + 20;
+	hintPos.y = rClient.top + sy - 20;
+
+	char primaryBuf[256];
+	char secondaryBuf[512];
+	BrushTool::getModeHintStrings(primaryBuf, sizeof(primaryBuf), secondaryBuf, sizeof(secondaryBuf));
+	
+	const char* firstLine = secondaryBuf && strlen(secondaryBuf) > 0 ? secondaryBuf : primaryBuf;
+	
+	Bool needUpdate = false;
+	if (modeInt != m_lastBrushMode) {
+		needUpdate = true;
+	} else if (::IsRectEmpty(&m_lastHintRect)) {
+		needUpdate = true;
+	} else {
+		Int dx = abs(hintPos.x - m_lastHintPos.x);
+		Int dy = abs(hintPos.y - m_lastHintPos.y);
+		if (dx > 15 || dy > 15) {
+			needUpdate = true;
+		}
+	}
+	
+	if (!firstLine || strlen(firstLine) == 0) {
+		if (!::IsRectEmpty(&m_lastHintRect)) {
+			::SetRectEmpty(&m_lastHintRect);
+			m_lastBrushMode = -1;
+			m_lastHintPos = CPoint(-10000, -10000);
+		}
+		return;
+	}
+
+	RECT hintRect;
+	SIZE textSize;
+	if (!needUpdate && !::IsRectEmpty(&m_lastHintRect)) {
+		hintPos = m_lastHintPos;
+		hintRect = m_lastHintRect;
+		if (m3DFont && !hdc) {
+			textSize.cx = (Int)strlen(firstLine) * 6;
```

Comment:
`strlen(firstLine)` keeps getting repeated multiple times which is unnecessary. Result should be cached.